### DISCUSSION TITLE
How to use `run_requests` and `dynamic_partitions_requests` with `SensorResult`

### DISCUSSION THREAD

**Discussion by jamiedemaria** [dagster-io/dagster#23248] [Q&A] [answered]
@jamiedemaria, I can't undesrstand how to construct run_requests and dynamic_partitions_requests for the SensorResult to return it from sensor?

_Originally posted by @andkuzmich in https://github.com/dagster-io/dagster/discussions/14334#discussioncomment-10142355_

**Comment by jamiedemaria** [comment, marked as answer]
here's a code snippet 

```python
dynamic_colors = DynamicPartitionsDefinition(name="dynamic_colors")

@asset(
    partitions_def=dynamic_colors,
)
def colors_asset(context: AssetExecutionContext):
    context.log.info(context.partition_key)


@sensor(asset_selection=[colors_asset])
def run_colors():
    colors = ["purple", "orange"]
    run_requests = []
    for color in colors:
        run_requests.append(RunRequest(run_key=color, partition_key=color))

    return SensorResult(
        run_requests=run_requests,
        dynamic_partitions_requests=[AddDynamicPartitionsRequest(partitions_def_name="dynamic_colors", partition_keys=colors)]
    )

defs = Definitions(assets=[colors_asset], sensors=[run_colors])
```

the [sensors concept page](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors) has some more information about run requests. You can learn more about dynamic partitions and `AddDynamicPartitionsRequest` [here](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitioning-assets#dynamically-partitioned-assets)

**Comment by andkuzmich** [comment]
@jamiedemaria, thank you for your reply!
Original question was about using MultiPartitionsDefinition combined with the DynamicPartitionsDefinition:

```
partitions_def = MultiPartitionsDefinition(
            {
                "time": DailyPartitionsDefinition(start_date="2022-01-01"),
                "color": dynamic_colors,
            }
        )
```
And my question was: what is the correct way to say Dagster that it should run a partition with the defined day and list of 'colours' in this case? The code snippet that you provided is about DynamicPartitions usage.

My problem is related to IoT where I have daily partitioned data but for different sensors. I have one asset, that collects sensors that was available for that day, than I have the sensor that monitors the asset materialization and it recieves the collected sensor names and then I hope to trigger the another asset materialization per time window per sensor name. I'm new with Dagster and I totally stuck with this.

**Comment by jamiedemaria** [comment]
ah, apologies for the confusion about that. If you want to make a `RunRequest` for a multi partition you would specify the partition you want the `RunRequest` to run using the `MultiPartitionKey` class

```python 
dynamic_colors = DynamicPartitionsDefinition(name="dynamic_colors")
partitions_def = MultiPartitionsDefinition(
            {
                "time": DailyPartitionsDefinition(start_date="2022-01-01"),
                "color": dynamic_colors,
            }
        )

@asset(
    partitions_def=partitions_def,
)
def colors_asset(context: AssetExecutionContext):
    context.log.info(context.partition_key)


@sensor(asset_selection=[colors_asset])
def run_colors():
    colors = ["purple", "orange"]
    run_requests = []
    for color in colors:
        run_requests.append(RunRequest(run_key=color, partition_key=MultiPartitionKey({"time": "2022-01-05", "color": color})))

    return SensorResult(
        run_requests=run_requests,
        dynamic_partitions_requests=[AddDynamicPartitionsRequest(partitions_def_name="dynamic_colors", partition_keys=colors)]
    )

defs = Definitions(assets=[colors_asset], sensors=[run_colors])
```

**Reply by gfaugere-chut** [reply]
In this example, you are using a static time in your `partition_key` - doesn't this matter? Shouldn't the date be dynamic too?

**Reply by gfaugere-chut** [reply]
Figured it out with: https://github.com/dagster-io/dagster/discussions/20508

Sensor is not returning run requests, only adding partitions to the partition store

A separate schedule creates the run requests 😊
